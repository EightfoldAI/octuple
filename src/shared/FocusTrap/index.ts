export * from './FocusTrap';
export * from './hooks/useFocusTrap';
