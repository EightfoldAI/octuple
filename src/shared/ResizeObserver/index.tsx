import React from 'react';
import toArray from '../toArray';
import SingleObserver from './SingleObserver';
import { Collection } from './Collection';

const INTERNAL_PREFIX_KEY = 'rc-observer-key';

export interface SizeInfo {
    width: number;
    height: number;
    offsetWidth: number;
    offsetHeight: number;
}

export type OnResize = (size: SizeInfo, element: HTMLElement) => void;

export interface ResizeObserverProps {
    /** Pass to ResizeObserver.Collection with additional data */
    data?: any;
    children:
        | React.ReactNode
        | ((ref: React.RefObject<any>) => React.ReactElement);
    disabled?: boolean;
    /** Trigger if element resized. Will always trigger when first time render. */
    onResize?: OnResize;
}

function ResizeObserver(props: ResizeObserverProps) {
    const { children } = props;
    const childNodes =
        typeof children === 'function' ? [children] : toArray(children);

    return childNodes.map((child: any, index: number) => {
        const key = child?.key || `${INTERNAL_PREFIX_KEY}-${index}`;
        return (
            <SingleObserver {...props} key={key}>
                {child}
            </SingleObserver>
        );
    }) as any as React.ReactElement;
}

ResizeObserver.Collection = Collection;

export default ResizeObserver;
