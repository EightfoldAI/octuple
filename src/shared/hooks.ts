import React, { useState, useEffect } from 'react';

export enum Breakpoints {
    Large = '(min-width: 1200px)',
    Medium = '(min-width: 900px)',
    Small = '(min-width: 600px)',
    XSmall = '(max-width: 599px)',
}

/**
 * A Hook to do CSS Media Queries in components.
 * @param breakpoint The mobile-first '(min-width: n)' threshold.
 * @returns a boolean determining if the threshold viewport size has been hit.
 */
export const useMatchMedia = (breakpoint: Breakpoints): boolean => {
    const [threshold, setThreshold]: [
        boolean,
        React.Dispatch<React.SetStateAction<boolean>>
    ] = useState(window.matchMedia(breakpoint).matches);
    useEffect((): void => {
        window
            .matchMedia(breakpoint)
            .addEventListener('change', (e) => setThreshold(e.matches));
        return window
            .matchMedia(breakpoint)
            .removeEventListener('change', (e) => setThreshold(e.matches));
    }, []);
    return threshold;
};
