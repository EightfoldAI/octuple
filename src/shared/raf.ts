let raf: (callback: FrameRequestCallback) => number = (
    callback: FrameRequestCallback
) => +setTimeout(callback, 16);
let caf: (num: number) => void = (num: number) => clearTimeout(num);

if (typeof window !== 'undefined' && 'requestAnimationFrame' in window) {
    raf = (callback: FrameRequestCallback) =>
        window.requestAnimationFrame(callback);
    caf = (handle: number) => window.cancelAnimationFrame(handle);
}

let rafUUID: number = 0;
const rafIds: Map<number, number> = new Map<number, number>();

const cleanup = (id: number) => {
    rafIds.delete(id);
};

export default function wrapperRaf(callback: () => void, times = 1): number {
    rafUUID += 1;
    const id: number = rafUUID;

    const callRef = (leftTimes: number): void => {
        if (leftTimes === 0) {
            // Clean up
            cleanup(id);

            // Trigger
            callback();
        } else {
            // Next raf
            const realId = raf(() => {
                callRef(leftTimes - 1);
            });

            // Bind real raf id
            rafIds.set(id, realId);
        }
    };

    callRef(times);

    return id;
}

wrapperRaf.cancel = (id: number): void => {
    const realId: number = rafIds.get(id);
    cleanup(realId);
    return caf(realId);
};
