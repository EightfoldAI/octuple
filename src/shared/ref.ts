import type * as React from 'react';
import useMemo from '../hooks/useMemo';

export function fillRef<T>(ref: React.Ref<T>, node: T) {
    if (typeof ref === 'function') {
        ref(node);
    } else if (typeof ref === 'object' && ref && 'current' in ref) {
        (ref as any).current = node;
    }
}

/**
 * Merge refs into one ref function to support ref passing.
 */
export function composeRef<T>(...refs: React.Ref<T>[]): React.Ref<T> {
    const refList = refs.filter((ref) => ref);
    if (refList.length <= 1) {
        return refList[0];
    }

    return (node: T) => {
        refs.forEach((ref) => {
            fillRef(ref, node);
        });
    };
}

export function useComposeRef<T>(...refs: React.Ref<T>[]): React.Ref<T> {
    return useMemo(
        () => composeRef(...refs),
        refs,
        (prev, next) =>
            prev.length === next.length &&
            prev.every((ref, i) => ref === next[i])
    );
}
