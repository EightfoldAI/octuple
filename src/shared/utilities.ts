/**
 *  Value represented as a generic (used by classNames).
 */
type Value = string | number | boolean | undefined | null;
/**
 *  Mapping that constructs a type with a set of properties K of type T (used by classNames).
 */
type Mapping = Record<string, unknown>;
/**
 *  ArgumentArray (used by classNames).
 */
interface ArgumentArray extends Array<Argument> {}
/**
 *  Argument represented as a Value, Mapping or ArgumentArray (used by classNames).
 */
type Argument = Value | Mapping | ArgumentArray;

/**
 *  Generates a string of class names.
 *  @param {ArgumentArray} args - ClassName input.
 *  @returns {string} - a concatenated string of class names.
 */
export function classNames(...args: ArgumentArray): string {
    const hasOwn: (v: PropertyKey) => boolean = {}.hasOwnProperty;
    let classes = [];
    for (let i: number = 0; i < args.length; i++) {
        const arg: any = args[i];
        if (!arg) continue;
        const argType = typeof arg;
        if (argType === 'string' || argType === 'number') {
            classes.push(arg);
        } else if (Array.isArray(arg)) {
            if (arg.length) {
                const inner = classNames.apply(null, arg);
                if (inner) {
                    classes.push(inner);
                }
            }
        } else if (argType === 'object') {
            if (arg.toString === Object.prototype.toString) {
                for (var key in arg) {
                    if (hasOwn.call(arg, key) && arg[key]) {
                        classes.push(key);
                    }
                }
            } else {
                classes.push(arg.toString());
            }
        }
    }
    return classes.join(' ');
}

/**
 *  RGB object type with red, green and blue components (used by invertForegroundColor).
 */
type RGB = {
    r: number;
    g: number;
    b: number;
};
/**
 *  RGB list (array) type with red, green and blue components (used by invertForegroundColor).
 */
type RgbArray = [number, number, number];
/**
 *  Hexadecimal representation of a color (used by invertForegroundColor).
 */
type HexColor = string;
/**
 *  Color represented as hexadecimal value or as RGB object or list (used by invertForegroundColor).
 */
type Color = RGB | RgbArray | HexColor;
/**
 *  Interface for defining black and white colors; used to amplify the contrast
 *  of the color inversion (used by invertForegroundColor).
 */
interface BlackWhite {
    black: HexColor;
    white: HexColor;
    threshold?: number;
}

/**
 *  Constants used by invertForegroundColor.
 */
const DEFAULT_THRESHOLD = Math.sqrt(1.05 * 0.05) - 0.05;
const RE_HEX = /^(?:[0-9a-f]{3}){1,2}$/i;
const DEFAULT_BW: BlackWhite = {
    black: '#000000',
    white: '#ffffff',
    threshold: DEFAULT_THRESHOLD,
};

/**
 *  Helper methods used by invertForegroundColor.
 */
function getLuminance(c: RgbArray): number {
    let i, x;
    const a = []; // so we don't mutate
    for (i = 0; i < c.length; i++) {
        x = c[i] / 255;
        a[i] = x <= 0.03928 ? x / 12.92 : Math.pow((x + 0.055) / 1.055, 2.4);
    }
    return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2];
}

function hexToRgbArray(hex: string): RgbArray {
    if (hex.slice(0, 1) === '#') hex = hex.slice(1);
    if (!RE_HEX.test(hex)) throw new Error(`Invalid HEX color: "${hex}"`);
    // normalize / convert 3-chars hex to 6-chars.
    if (hex.length === 3) {
        hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
    }
    return [
        parseInt(hex.slice(0, 2), 16), // r
        parseInt(hex.slice(2, 4), 16), // g
        parseInt(hex.slice(4, 6), 16), // b
    ];
}

function toRgbArray(c: Color): RgbArray {
    if (!c) throw new Error('Invalid color value');
    if (Array.isArray(c)) return c as RgbArray;
    return typeof c === 'string' ? hexToRgbArray(c) : [c.r, c.g, c.b];
}

function invertToBW(color: RgbArray): HexColor {
    const options = DEFAULT_BW;
    return getLuminance(color) > options.threshold
        ? options.black
        : options.white;
}

/**
 *  Use only the needed code from https://github.com/onury/invert-color/blob/master/src/invert.ts
 *  Generates inverted (opposite) version of the given color.
 *  @param {Color} color - Color to be inverted.
 *  @returns {HexColor} - Hexadecimal representation of the inverted color.
 */
export function invertForegroundColor(color: Color): HexColor {
    color = toRgbArray(color);
    return invertToBW(color) as HexColor;
}

/**
 * https://dev.to/bwca/create-a-debounce-function-from-scratch-in-typescript-560m
 * @param fn Accepted argument and return, a function.
 * @param ms The delay time.
 * @returns a debounced function that resolves to a value after certain amount of time. Or never resolves if terminated.
 */
export function debounce<A = unknown, R = void>(
    fn: (args: A) => R,
    ms: number
): [(args: A) => Promise<R>, () => void] {
    let timer: NodeJS.Timeout;

    const debouncedFunc = (args: A): Promise<R> =>
        new Promise((resolve) => {
            if (timer) {
                clearTimeout(timer);
            }

            timer = setTimeout(() => {
                resolve(fn(args));
            }, ms);
        });

    const teardown = () => clearTimeout(timer);

    return [debouncedFunc, teardown];
}

/**
 * Generates a simple unique id
 * @param prefix The sting prefix
 */
export const uniqueId = ((): ((prefix: string) => string) => {
    let counter: number = 0;
    return (prefix: string): string => `${prefix}${++counter}`;
})();
