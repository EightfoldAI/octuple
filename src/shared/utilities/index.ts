export * from './addEventListenerWrapper';
export * from './canUseDom';
export * from './conditionalWrapper';
export * from './contains';
export * from './css';
export * from './debounce';
export * from './domWrapper';
export * from './eventKeys';
export * from './findDOMNode';
export * from './flexGapSupported';
export * from './generateId';
export * from './get';
export * from './getScroll';
export * from './getScrollBarSize';
export * from './isNumeric';
export * from './isVisible';
export * from './mergeClasses';
export * from './mobileDetector';
export * from './motion';
export { default as collapseMotion } from './motion';
export * from './omit';
export * from './overflow';
export * from './pickAttrs';
export * from './raf';
export * from './reactNode';
export * from './ref';
export * from './responsiveObserve';
export { default as scrollTo } from './scrollTo';
export * from './set';
export * from './stopPropagation';
export * from './styleChecker';
export { default as toArray } from './toArray';
export * from './types';
export * from './uniqueId';
export * from './utilities.types';
export { default as visuallyHidden } from './visuallyHidden';
export * from './statusUtils';
export * from './warning';
export * from './resolveOnChange';
