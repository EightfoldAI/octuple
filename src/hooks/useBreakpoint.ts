import { useEffect, useRef } from 'react';
import { useForceUpdate } from './useForceUpdate';
import type { Breakpoint, ScreenMap } from '../shared/utilities';
import { responsiveObserve } from '../shared/utilities';

export const useBreakpoint: (refreshOnChange?: boolean) => ScreenMap = (
  refreshOnChange: boolean = true
): ScreenMap => {
  const screensRef: React.MutableRefObject<
    Partial<Record<Breakpoint, boolean>>
  > = useRef<ScreenMap>({});
  const forceUpdate: React.DispatchWithoutAction = useForceUpdate();

  useEffect(() => {
    const token: number = responsiveObserve.subscribe((supportScreens) => {
      screensRef.current = supportScreens;
      if (refreshOnChange) {
        forceUpdate();
      }
    });

    return () => responsiveObserve.unsubscribe(token);
  }, []);

  return screensRef.current;
};
