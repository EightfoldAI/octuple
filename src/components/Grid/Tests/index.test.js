import React from 'react';
import Enzyme, { mount, render } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { act } from 'react-dom/test-utils';
import { Col, Row } from '..';
import { responsiveObserve } from '../../../shared/utilities';
import useBreakpoint from '../../../hooks/useBreakpoint';
import { render as libRender } from '@testing-library/react';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia;

jest.mock('../../../hooks/useBreakpoint', () => () => {
  return { xs: true, sm: false, md: false, lg: false, xl: false };
});

describe('Grid', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('should render Col', () => {
    const wrapper = render(<Col span={2} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render Row', () => {
    const wrapper = render(<Row />);
    expect(wrapper).toMatchSnapshot();
  });

  test('when typeof gutter is object', () => {
    const { container } = libRender(<Row gutter={{ xs: 8 }} />);
    expect(container.querySelector('.row').style.marginLeft).toEqual('-4px');
    expect(container.querySelector('.row').style.marginRight).toEqual('-4px');
  });

  test('when typeof gutter is object array', () => {
    const { container } = libRender(<Row gutter={[{ xs: 8 }, { xs: 8 }]} />);
    expect(container.querySelector('.row').style.marginLeft).toEqual('-4px');
    expect(container.querySelector('.row').style.marginRight).toEqual('-4px');
  });

  test('when typeof gutter is object array in large screen', () => {
    const wrapper = render(
      <Row
        gutter={[
          { xs: 8, sm: 16, md: 24, lg: 32 },
          { xs: 8, sm: 16, md: 24, lg: 100 },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('renders wrapped Col correctly', () => {
    const MyCol = () => <Col span={12} />;
    const wrapper = render(
      <Row gutter={20}>
        <div>
          <Col span={12} />
        </div>
        <MyCol />
      </Row>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('responsiveObserve.unsubscribe should be called when unmounted', () => {
    const Unmount = jest.spyOn(responsiveObserve, 'unsubscribe');
    const wrapper = mount(<Row gutter={{ xs: 20 }} />);
    act(() => {
      wrapper.unmount();
    });
    expect(Unmount).toHaveBeenCalled();
  });

  test('should work correct when gutter is object', () => {
    const { container } = libRender(<Row gutter={{ xs: 20 }} />);
    expect(container.querySelector('.row').style.marginLeft).toEqual('-10px');
    expect(container.querySelector('.row').style.marginRight).toEqual('-10px');
  });

  test('should work current when gutter is array', () => {
    const wrapper = mount(<Row gutter={[16, 20]} />);
    expect(wrapper.find('div').prop('style')).toEqual({
      marginLeft: -8,
      marginRight: -8,
      marginTop: -10,
      marginBottom: -10,
    });
  });

  // By jsdom mock, actual jsdom not implemented matchMedia
  // https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
  test('should work with useBreakpoint', () => {
    const screens = useBreakpoint();
    function Demo() {
      return JSON.stringify(screens);
    }
    const wrapper = mount(<Demo />);

    expect(wrapper.text()).toEqual(
      JSON.stringify({
        xs: true,
        sm: false,
        md: false,
        lg: false,
        xl: false,
      })
    );
  });
});
