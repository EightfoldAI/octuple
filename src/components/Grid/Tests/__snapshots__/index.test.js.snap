// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders wrapped Col correctly 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "row",
      "style": "margin-left:-10px;margin-right:-10px",
    },
    "children": Array [
      Node {
        "attribs": Object {},
        "children": Array [
          Node {
            "attribs": Object {
              "class": "col col-12",
              "style": "padding-left:10px;padding-right:10px",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "col col-12",
            "style": "padding-left:10px;padding-right:10px",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Node {
        "attribs": Object {
          "class": "col col-12",
          "style": "padding-left:10px;padding-right:10px",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {
                "class": "col col-12",
                "style": "padding-left:10px;padding-right:10px",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "style": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Grid should render Col 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "col col-2",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Grid should render Row 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "row",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Grid when typeof gutter is object array in large screen 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "row",
      "style": "margin-left:-16px;margin-right:-16px;margin-top:-50px;margin-bottom:-50px",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
